data/transactions/logic/evalAppTxn_test.go

```diff
@@ -564,19 +632,1094 @@ txn Sender; itxn_field Receiver;
+   itxn CreatedApplicationID; app_params_get AppAddress; assert; itxn_field Receiver
+   int 10;                    itxn_field Amount
+  itxn_submit
+
+  int 1
+`
+
+	ep, tx, ledger := MakeSampleEnv()
+	ledger.NewApp(tx.Receiver, 888, basics.AppParams{})
+	ledger.NewAccount(appAddr(888), 10*MakeTestProto().MinTxnFee)
+	TestApp(t, createAndPay, ep)
+
+	// This test is impossible because CreatedResourcesVersion is also when
+	// inner txns could make apps.
+	// ep.Proto = MakeTestProtoV(CreatedResourcesVersion - 1)
+	// TestApp(t, createAndPay, ep, "invalid Address reference")
+}
+
+// TestInnerGaid ensures there's no confusion over the tracking of ids
+// across multiple inner transaction groups
+func TestInnerGaid(t *testing.T) {
+	ep, tx, ledger := MakeSampleEnv()
+	ep.Proto.MaxInnerTransactions = 100
+	// App to log the aid of slot[apparg[0]]
+	logGaid := TestProg(t, `txn ApplicationArgs 0; btoi; gaids; itob; log; int 1`, AssemblerMaxVersion)
+	ledger.NewApp(tx.Receiver, 222, basics.AppParams{
+		ApprovalProgram: logGaid.Program,
+	})
+
+	ledger.NewApp(tx.Receiver, 888, basics.AppParams{})
+	ledger.NewAccount(appAddr(888), 50_000)
+	tx.ForeignApps = []basics.AppIndex{basics.AppIndex(222)}
+	TestApp(t, `itxn_begin
+int acfg;    itxn_field TypeEnum
+itxn_next
+int pay;      itxn_field TypeEnum
+txn Sender;   itxn_field Receiver
+itxn_next
+int appl;    itxn_field TypeEnum
+int 222;     itxn_field ApplicationID
+int 0; itob; itxn_field ApplicationArgs
+itxn_submit
+itxn Logs 0
+btoi
+int 5000
+==
+assert
+
+// Swap the pay and acfg, ensure gaid 1 works instead
+itxn_begin
+int pay;      itxn_field TypeEnum
+txn Sender;   itxn_field Receiver
+itxn_next
+int acfg;    itxn_field TypeEnum
+itxn_next
+int appl;    itxn_field TypeEnum
+int 222;     itxn_field ApplicationID
+int 1; itob; itxn_field ApplicationArgs
+itxn_submit
+itxn Logs 0
+btoi
+int 5001
+==
+assert
+
+
+int 1
+`, ep)
+
+	// Nearly identical, but ensures that gaid 0 FAILS in the second group
+	TestApp(t, `itxn_begin
+int acfg;    itxn_field TypeEnum
+itxn_next
+int pay;      itxn_field TypeEnum
+txn Sender;   itxn_field Receiver
+itxn_next
+int appl;    itxn_field TypeEnum
+int 222;     itxn_field ApplicationID
+int 0; itob; itxn_field ApplicationArgs
+itxn_submit
+itxn Logs 0
+btoi
+int 5000
+==
+assert
+
+// Swap the pay and acfg, ensure gaid 1 works instead
+itxn_begin
+int pay;      itxn_field TypeEnum
+txn Sender;   itxn_field Receiver
+itxn_next
+int acfg;    itxn_field TypeEnum
+itxn_next
+int appl;    itxn_field TypeEnum
+int 222;     itxn_field ApplicationID
+int 0; itob; itxn_field ApplicationArgs
+itxn_submit
+itxn Logs 0
+btoi
+int 5001
+==
+assert
+
+
+int 1
+`, ep, "assert failed")
+
+}
+
+// TestInnerCallDepth ensures that inner calls are limited in depth
+func TestInnerCallDepth(t *testing.T) {
+	t.Parallel()
+
+	ep, tx, ledger := MakeSampleEnv()
+	// Allow a lot to make the test viable
+	ep.Proto.MaxAppTxnForeignApps = 50
```

Comment:
Before we decided to do the limit on maxAppCallDepth, this code was written to go much deeper. So it seems good to leave the test in that makes it so that it _could_ go deep, and ensure that it can't.

Reply:
If we don't allow a bunch of ForeignApps, we can't construct a txn that will even get that deep. We have to call at least 8 different apps to check the depth limit, and the default ForeignApps limit is smaller.